### ISSUE TITLE
Serial to SAPI or ESPeak for blind DOS users

### ISSUE THREAD

**Issue by blindphoneman** [joncampbell123/dosbox-x#1919]
Hi:

I am a blind computer user who started using DOS back in the '80's.

Due to projects like this one, interest in DOS by the blind is being revived.

Currently, we can use any of the emulators with a null-modem emulator (com0com), which then communicates with an emulator to simulate an external serial speech synthesizer.

This can be tricky to set up, though, and it occurs to me that, perhaps a new use could be made of DOSBox-X's serial port system.

Instead of using 3 different programs, could some kind of module be incorporated into DOSBox-X that could be addressed something like this:

serial1   = speech

When DOSBox-X sees this, it could redirect output to a serial port to the SAPI speech routines in Windows. I believe Linux also offers this, eand MAC-OS has speech as well, so it could work across platforms.

The ESpeak synthesizer could also be used, as it is lightweight, and I think is open-source.

Using something like this, a blind person could load his favorite screen-reader, which runs as a tsr intercepting DOS output and putting it out the serial port, with some control codes, when the program wants to change the pitch or speed of the voice.

Though the screen-readers are not free, as such, most of us still have our old files around,.

As for control codes,  the minimum might be to use CTRL+X to silence the speech.  But, it is possible to have some codes like CTRL+A, a number, and a letter to change, say, the speed. For example:

^A5S

to set the speed to 5.  Some synthesizers used numbers from 1 to 9, but, others may have had finer adjustments.  I believe, there was also CTRL+a, a number from 1 to 9, followed by a letter P to change the pitch.  This is used a lot to identify capital letters when screen is being reviewed character-by-character.  The pitch can go up to signify a capital letter, then back to normal for a lower-case letter.

The reason I see a need for this is that the setup of the current arrangement involves configuring, not only DOSBox-X, but, the COM0COM package, which may require turning off or modifying SecureBoot settings in UEFI firmware, something that a blind person may find difficult, if not impossible without sighted assistance.

This way, the blind person would only need to set up the DOSBox-X config file, then run DOSBox-X and the SAPI speech would take over.  No need to run 2 other programs and no need to get sighted assistance for fiddling with UEFI settings.  Again, the minimum of a CTRL+X would be needed to silence speech, but, it would be nice to have the others, at least, the ones to change pitch and speed.  Don't know about a buffer-size limit, but, some speech synthesizers do send X-Off and X-On when their buffers are too full.

It is also possible to use some kind of 2-way communication, called "indexing" for the speech synthesizer to let the screen-reader know what it is speaking,  This makes it possible for the screen-reader to be automatically reading text, and, when the alt or ctrl key is pressed, the screen-reader knows where the person has left off, but, that's probably in the realm of wishful thinking.

I would be willing to help test this and find out what codes are needed.  Think I can dig up some manuals or something to get specifics on the codes.

Thanks for your attention.

**Comment by blindphoneman** [comment]
Hi again:

Just wanted to make one more clarification.

When I mention using control codes, like ctrl+a to begin voice adjustment or ctrl+x to silence speech (clearing the buffers), these are not typed by the user, but, sent to the serial port (or isa bus for internal speech cards) by the screen-reader program itself.

I should mention that some speech devices may also use hardware handshaking, but, not sure which do.  If there's interest, I'll find out.

**Comment by Jookia** [comment]
Interesting idea! It sounds like what you're describing is emulation of a hardware speech synthesizer, except using a software synthesizer as the actual hardware. Could you dig up some information on these types of protocols? Or an open source screen reader?

**Comment by joncampbell123** [comment]
On Windows it could be a serial device that sends it's output to Microsoft's Speech API.

https://docs.microsoft.com/en-us/previous-versions/windows/desktop/ms720571(v=vs.85)

**Comment by joncampbell123** [comment]
A little more digging around could support SAPI 4 which has all the classic text to speech voices you used to hear all the time in the early 2000s. I think I have that SDK around somewhere as well.

**Comment by joncampbell123** [comment]
For Linux, there are several options.

https://blog.michaelamerz.com/wordpress/the-state-of-the-art-linux-text-to-speech-tts/

**Comment by joncampbell123** [comment]
Finally for Mac OS X, there is a speech API (and speech recognition API) that is *gasp* actually reasonably documented!

https://developer.apple.com/documentation/appkit/nsspeechsynthesizer#//apple_ref/doc/uid/TP40004114

**Comment by joncampbell123** [comment]
I know from programming experience with Microsoft's speech API that it does allow you to direct the TTS output to whatever you want, it doesn't have to be the sound card, so it could be directed into DOSBox-X's audio mixing system and recorded with everything else.

**Comment by blindphoneman** [comment]
Windows doesn’t need data sent over a serial port.  You could also use Espeak so as not to need to use SAPI.  The 2 VBNS projects (one of which I referenced earlier) emulates that.  It would seem like you would eliminate a middleman if you went right from DOSBox-X either to SAPI or to ESpeak.

**Comment by blindphoneman** [comment]
Hadn’t thought of that.

 

I’m not particular, but, wouldn’t that be more complicated and slow things down inside DOSBox-X?

 

Using SAPI speech and using the Windows soundcard is just like when we used an external synth, and we were ok with that.

 

If it didn’t slow things down or interfere with the DOS system itself, wouldn’t hurt.

 

Someone asked me if I used Provox.  It is useable, but, not most of us blind users’ favorite.  However, as long as data can be directed as if it is coming out a serial port and SAPI speaks it, it wouldn’t matter which screen reader was used.  They all work well enough with serial synths.

 

I don’t know if any of the manufacturers of ISA synthesizers could be found, but, those just involve using an ISA slot.  Can’t remember which interrupt my internal DoubleTalk used, but, it was made by RC Systems.  Their products, though, are not open-source, though.

 

Again, the VBNS (virtual Braille ‘n Speak) programs work, it would just be nice not to need a null-modem bridge between DOSBox-X and the VBNS program.

**Comment by blindphoneman** [comment]
Yep, Apple have made sure anyone who wants to have accessibility has the API’s for it.

 

I think data can be processed by VoiceOver, or just sent to the speech api itself.

 

So, the Windows speech api is not as well documented?  I’d wondered why more was not done in the open-source community.

**Comment by joncampbell123** [comment]
Routing TTS through DOSBox-X would be no different than Linux builds where MIDI output goes through Fluidsynth and directed through the DOSBox-X mixer.

Also knowing Microsoft, they probably have the TTS engine run it's synthesis in another thread anyway, and even if it didn't, the old SAPI engines hardly took any CPU even on early 2000s hardware.

**Comment by blindphoneman** [comment]
I think SAPI 4 might still be workable in Windows 10 even.  I think I’ve got some SAPI 4 voices around as well.

**Comment by joncampbell123** [comment]
@blindphoneman Exactly. I doubt there will ever be a 64-bit binary of them, but the old 32-bit binaries should do just fine.

**Comment by blindphoneman** [comment]
Yes.  The 32-bit ones work well enough.  I still have several, including a resource-hog, a couple of the AT&T natural voices.  I’d hate to try and use those.

Again, ESpeak is even more lightweight and isn’t it open-source?  That’s what ORCA uses in Linux.


Sent from Mail for Windows 10

**Comment by Jookia** [comment]
Without any documentation of serial protocols, emulating whatever protocol DOS-based screen readers expect isn't possible. An open source emulator would be really helpful.

I found this DECTalk manual that specifies the protocol a bit: https://archive.org/details/EK-DTC01-OM-002_Owners_Manual
These DoubleTalk manuals are great: http://www.cs.uni.edu/~mccormic/4740/documentation/DoubleTalk_Tools.pdf http://www.cs.uni.edu/~mccormic/4740/documentation/DoubleTallk_User_Manual.pdf
The most I could find is

**Comment by blindphoneman** [comment]
Hi:

My mistake:  That’s where I remembered CTRL+a from.

DoubleTalk uses a ctrl+a followed by a number and a letter to set such things as pitch and speed.

If those 2 could be emulated, all a user would have to do is specify a doubletalk synthesizer to any screen reader during setup or start of the tsr and it would use those two most often.  The emulator could ignore others.  I don’t know if those are considered proprietary or not.  The Echo uses similar commands, but, with a ctrl+e instead of ctrl+a and there are only 2 rates with at least some of the echos, ctrl+e followed by c for fast speech and ctrl+e followed by e for slow (extended) speech.  The only other major thing might be a ctrl+x which, when received, would stop speech and ready the synth for new input.  If the buffer were too full, maybe using rts/cts handshaking to stop sending until there is room in the buffer.

However, could the code in the VBNS emulator at:

https://github.com/sukiletxe/vbns-ao2

or the eSpeak one at:

https://github.com/sukiletxe/vbns-eSpeak

be somehow incorporated, eliminating the need for the external serial ports and such?

Those have been around for a while and no apparent issues.

I’m just hoping we can get around using kernel-mode drivers as those can be hard for newbies to install and some folks might not want to disable SecureBoot, thogh I don’t mind that much.


Sent from Mail for Windows 10

**Comment by Jookia** [comment]
As far as I can tell, the VBNS code is proprietary so using that as a basis for any emulator code is kind of shaky territory.
Starting from scratch seems like the only option here unless there's other emulators out tehre.

**Comment by blindphoneman** [comment]
Well, what about just starting with sort-of a generic thing.

You could just send everything to the synth, maybe ignoring strings that start with a control character, an optional number and letter, just using control+x to empty the speech buffers and silence speech.  CTRL+X is used by several manufacturers, so you should be on safe ground there.

The user could adjust default voice, speed,  pitch and such in the sapi control panel ahead of time.  Using CTS+RTS to stop sending when the buffer is full should also be safe, since that is pretty-much standard.

Those should be safe.


Sent from Mail for Windows 10

**Comment by Jookia** [comment]
Control chracters vary across devices wouldn't they? I guess if we had a screen reader to test with we could just dump and look at the output. Also by ctrl-x do you mean the button the user presses in DOS? Shouldn't the DOS screen reader handle that?

Ignoring voice/speed/pitch etc in the control panel might work. Having a generic synth driver that works would be better than nothing I suppose, and could be a good way to start.

**Comment by blindphoneman** [comment]
I forgot that, maybe a parameter could be added to specify a default voice if using eSpeak, say,

Serial1 = speech voice:en-us

With also something to set rate and such.

Not as flexible as using a set of synth protocols with the screen reader, but, it might be workable that way if using eSpeak.



Sent from Mail for Windows 10

**Comment by Jookia** [comment]
Ah, I see based on your description. If I do implement this I won't be looking at the code for VBNS due to its lack of license.

**Comment by blindphoneman** [comment]
I didn't know the lack of license was an issue.  I figured if the vbns emulator was on Github it was fair game.

So, right now, the thing I am sure about is that several unrelated speech devices use ctrl+x to silence speech.  So, I think most of us can live with setting up speech parameters outside of DOSBox-X or some kind of initial setup on launch.  Most of the screen readers do their own thing with pronunciation and punctuation.  Many do raise the pitch of the voice temporarily when giving typing feedback to let the user know they have typed a capital letter, but, that particular thing is not implemented in any of the VBNS code for either eSpeak or SAPI.

Here's an idea:

Could the CTRL+X thing be put in, and some kind of placeholders for adding control codes for pitch, and speed change for future implementation?  If they were commented, maybe I could even add my own codes and recompile for my personal use to test, then try to make sure that licensing issues wouldn't be encountered?  If I knew where to insert the strings that were needed to be acted upon, I bet I could put them in.

Of course, I have no idea what I need to actually compile a Windows binary myself, but, I would sure be willing to try.

**Comment by Jookia** [comment]
Yeah, licensing is complicated.

I like the idea of adding a serial port speech emulator that will read out stuff sent in, cancel when it gets ^X, and ignore commands for pronounciation and other modes.

I'm currently working on a PR already, but I'll try and make this my next DOSBox-X hack. :)

**Comment by blindphoneman** [comment]
Hi:

 

That’s fantastic.

 

Just for variation, I do have a demo of another screen-reader and a shareware one, for which I found the registered version on vetusware.  It appeared to have uploaded by the author, and it has a serial number of 00000 or something like that, suggesting either the author did it or it was hacked.  I also have the original shareware files as well, but they have a built-in delay when the program first starts.

 

If you’d like any of that to get a feel for what it does, let me know.

 

Thanks!

**Comment by Jookia** [comment]
I'm not going to get to this for a while, but I'd really appreciate the following:

- Any guides on how to set up whatever screen reader setup using DOSBox
  that works, maybe like how you do now
- Any links to screen readers, guides, etc

The biggest hurdle to getting support for a feature like this is getting a working environment for testing or a control that we know works. Putting together a list of tools, guides, and some instructions from what you know would be really helpful.

As a reminder for myself later, I do want to scour the net for:

- Guides on hardware synthesizers used in the DOS era
- DOS screen readers
- Emulated hardware synthesizers for Windows

Some useful sites I've found so far:
https://www.oocities.org/hotsprings/villa/6113/s_tts1.htm
https://www.vogons.org/viewtopic.php?t=42012
https://allinaccess.com/happ/
https://nvda.groups.io/g/nvda/topic/screen_reader_under_dos/24959034?p=
http://www.nfbnet.org/download/blind.htm

The only open source DOS screen reader I know of is Provox. I did some
hunting and found the latest source code is here:
https://web.archive.org/web/20150310023633/http://www.hallenbeck.ftml.net/software.html
The file is provox7.zip with md5sum efa63e1ab90da3675e605be76629c0f9

**Comment by blindphoneman** [comment]
Hi:

 

I think probably the most thorough example would be what inspired me to ask about this in the first place.

 

Here is a set of stuff put together based on an earlier version of DOSBox-DAUM.  It’s simply called “Talking DOSBox.  It has full instructions right on the page at:

 

http://batsupport.com/unsupported/dosbox/

 

You’ll see a file called “talking_dosbox.zip” which can simply be unzipped to wherever you want.

 

The user needs to install Com0Com, and set it up for COM8 and COM9 as the port pair, not sure why this was picked.  Once this is done, the included startbns batch file can be run at the command prompt.  Several screen-readers are set up ready to go, and their included documentation.

 

A word about the screen-readers included.

 

The ASAP screen reader uses serial output, using the Com0Com null-modem emulator along with the VBNS for speech.  It provides the smoothest experience.

 

The JAWS and Flipper screen readers use the old SBTalker from the Sound Blaster, however, they can be made to use serial output when installed correctly, and this works better.  Speech with the Sound Blaster’s SBTalker driver needs to have cycles set to max and can be flaky.  Also, it was not used back in the day because it didn’t work with any other serial port due to interrupt issues.

 

My hope is to eliminate the need for the Com0Com, so as not to need kernel-mode drivers, and to eliminate the need for the VBNS program.

 

The developer of AppleWin has included output to SAPI speech, but, when you turn it on, it talks too much in some circumstances.  Keeping things looking like serial output, so as to use a screen reader would be the best setup.

 

Let me know if this approaches what you’re looking for.

 

Thanks!

**Comment by blindphoneman** [comment]
Hi:

 

I will also zip up my DOSBOX-X folder with my config file and working screen readers.

 

Documentation for each is included.

 

My setup can be run by just installing Com0Com, because it uses the default of com3 for the VBNS emulator and com4 for DOSBox-X.

 

Just unzip it somewhere, cd into “mydosbox-x” and type startbns.

 

If all is set up correctly, it will speak “Welcome to DOSBOX-X” and you can type one of the commands for the four screen readers there.

 

1 <https://www.dropbox.com/s/c2e6jun5i8g223k/myDOSBox-X.zip?dl=0>

**Comment by blindphoneman** [comment]
Hi:

 

I made some changes to the personal copy called “mydosbox-x” and rezipped it.  

 

I included most of the stuff from the Talking DosBox file, but, without the SoundBlaster screen reader stuff, so all the five screen readers use com1.

 

Installing Com0Com out of the box should work.  Here’s the new link

 

https://www.dropbox.com/s/a61hd8lgxama0ad/myDOSBox-X.zip?dl=1

 

This way, you can run anything there and see how it works.

**Comment by Jookia** [comment]
I just want to pop in and say no I haven't forgotten about this and I'm not going to let this issue get ignored like most accessibility bugs will, I'm still stuck in my PR over at #1863 which is already enough work. I'm *incredibly* interested in wrapping my head around this and trying to implement some of it.

**Comment by blindphoneman** [comment]
Hi: No problem.  Thanks!

**Comment by FelixGruetzmacher** [comment]
As a blind DOSBox-X user I am delighted to read that this won't eventually get ignored like most accessibility issues. Any news on this, for that matter?

**Comment by Jookia** [comment]
Hi there, I've been kind of ignoring this as I haven't contributed to DOSBox-X in a long time, in part because I said I'd do this. I've also managed to have a complete mental breakdown and burnout. 😅 But maybe this can be the year of me fixing this issue huh?

Felix (and anyone else), would you be willing to work with me on testing development builds I make?

**Comment by FelixGruetzmacher** [comment]
Absolutely, do count me in where testing---and to some degree development
in my spare time---is concerned! But no issue or bug is as important as
your mental health.

**Comment by Jookia** [comment]
Do you have an email/IRC/Discord/Matrix/something I can message you to talk in real time? Not to discuss things that would be here but to avoid cluttering up the issue here.

**Comment by blindphoneman** [comment]
Hi:

I also would be happy to test anything you come up with.

Already Emailed you.

No hurry--just take care of yourself.

**Comment by Jookia** [comment]
Hi there, I did start on working on a DECTalk serial protocol to SAPI bridge that could be plugged in to DOSBox-X or QEMU, but after a lot of frustration with modern technology and its approach to disability I've decided to abandon the effort and won't be completing this issue. Sorry.

**Comment by blindphoneman** [comment]
Hi:

 

That’s ok.

 

Thanks for all the effort anyway.

 

I’m still working quite nicely with the current system I have been using.

 

Thanks again.

 

Take care!
